### PR TITLE
FFI bindings for Language and Region Displaynames.

### PR DESCRIPTION
**Pull request by snktd**

### CODE REVIEW

File: ffi/diplomat/src/displaynames.rs
Comment (Manishearth) [inline ffi/diplomat/src/displaynames.rs:49]: nit: you should be able to use `#[diplomat::enum_convert]` here, then you don't need the `From` impl

(you still need it for `ICU4XStyle` unfortunately because of the `Auto`)

File: ffi/diplomat/src/displaynames.rs
Comment (Manishearth) [inline ffi/diplomat/src/displaynames.rs:57]: nit: also here

**Comment by Manishearth** [comment]
We should probably hold off on doing this until https://github.com/unicode-org/icu4x/pull/3190 lands?

Though we could land this and Rob could fix up the FFI in his PR

File: ffi/diplomat/src/displaynames.rs
Comment (snktd) [reply ffi/diplomat/src/displaynames.rs:50]: Done.

File: ffi/diplomat/src/displaynames.rs
Comment (snktd) [reply ffi/diplomat/src/displaynames.rs:60]: Done.

**Comment by snktd** [comment]
> We should probably hold off on doing this until #3190 lands?
> 
> Though we could land this and Rob could fix up the FFI in his PR

I am fine either way. @robertbastian ?

**Comment by Manishearth** [comment]
We decided to land this first

File: ffi/gn/.cargo/config
Comment (sffc) [inline ffi/gn/.cargo/config:10]: There should be no changes in this file

File: ffi/diplomat/src/displaynames.rs
Comment (sffc) [inline ffi/diplomat/src/displaynames.rs:41]: Issue: Please give this a more specific name like `ICU4XDisplayNamesStyle`

File: ffi/diplomat/src/displaynames.rs
Comment (sffc) [inline ffi/diplomat/src/displaynames.rs:52]: `ICU4XDisplayNamesFallback`

File: ffi/diplomat/src/displaynames.rs
Comment (sffc) [inline ffi/diplomat/src/displaynames.rs:84]: Here and below: Please document that the function returns an empty string in place of the None value.

**Review by sffc — CHANGES_REQUESTED** [review]

**Comment by robertbastian** [comment]
Is display names component stable enough to be unconditionally included in `icu_capi`? Maybe we should do #2602 first.

**Comment by sffc** [comment]
We have precedent for unstable components being included in icu_capi and I think this is basically stable enough for that

**Comment by Manishearth** [comment]
I think the only thing left is resolving Shane's comments, I'm ready to merge this once that's done

File: ffi/gn/.cargo/config
Comment (snktd) [reply ffi/gn/.cargo/config]: Ah not sure where this change come from. Removed now.

File: ffi/diplomat/src/displaynames.rs
Comment (snktd) [reply ffi/diplomat/src/displaynames.rs]: Done.

File: ffi/diplomat/src/displaynames.rs
Comment (snktd) [reply ffi/diplomat/src/displaynames.rs]: Done.

File: ffi/diplomat/src/displaynames.rs
Comment (snktd) [reply ffi/diplomat/src/displaynames.rs:84]: Yep done.

Comment (Manishearth) [reply ffi/diplomat/src/displaynames.rs:84]: Please use doc comments (three slashes)

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
This is blocked on https://github.com/unicode-org/icu4x/pull/3215 but you can feel free to include that commit in this PR

File: ffi/diplomat/src/displaynames.rs
Comment (snktd) [reply ffi/diplomat/src/displaynames.rs]: Done and regenerated ffi bindings to make sure that the comment is visible.

File: ffi/diplomat/src/displaynames.rs
Comment (sffc) [inline ffi/diplomat/src/displaynames.rs:60]: Thought: I wonder if this enum is sufficiently namespaced. It's probably fine.

**Review by sffc — APPROVED** [review]

File: ffi/diplomat/src/displaynames.rs
Comment (Manishearth) [reply ffi/diplomat/src/displaynames.rs:60]: I was also wondering that and came to the same conclusion

Comment (snktd) [reply ffi/diplomat/src/displaynames.rs:60]: Do you mean the enum name should be starting with ICU4XDisplayNames?

File: ffi/diplomat/src/displaynames.rs
Comment (sffc) [inline ffi/diplomat/src/displaynames.rs:37]: Actually, if you're willing to make more changes, I might prefer

```
    pub struct ICU4XDisplayNamesOptions {
        /// The optional formatting style to use for display name.
        pub length: ICU4XDisplayNamesLength,
        /// The fallback return when the system does not have the
        /// requested display name, defaults to "code".
        pub fallback: ICU4XDisplayNamesFallback,
        /// The language display kind, defaults to "dialect".
        pub language_style: ICU4XDisplayNamesLanguageStyle,
    }
```

and the Rust types should probably take those names, too.

This isn't necessarily consistent with 402 but we've had the discussion about "style" versus "length" before and I think we decided that ICU4X should be self-consistent.

**Pull request merged**
